ios/objc/src/TorchTensor.mm

```diff
@@ -0,0 +1,196 @@
+#import "TorchTensor.h"
+#import "TorchTensor+Internal.h"
+#import <Pytorch/Pytorch.h>
+
+#define CHECK_IMPL(x) NSCAssert(x!=nil,@"impl is nil!");
+#define CHECK_IMPL_(x) \
+    CHECK_IMPL(x) \
+    if (!x) { return nil; }
+
+#define DEFINE_TENSOR_TYPES(_) \
+    _(Byte) \
+    _(Int) \
+    _(Float) \
+    _(Long) \
+    _(Undefined)
+
+#define DEFINE_TENSOR_SCALAR_TYPES(_) \
+    _(Int) \
+    _(Float) \
+    _(Long) \
+
+static inline c10::ScalarType scalarTypeFromTensorType(TorchTensorType type) {
+    switch(type){
+#define DEFINE_CASE(x) case TorchTensorType##x: return c10::ScalarType::x;
+        DEFINE_TENSOR_TYPES(DEFINE_CASE)
+#undef DEFINE_CASE
+    }
+    return c10::ScalarType::Undefined;
+}
+
+static inline TorchTensorType tensorTypeFromScalarType(c10::ScalarType type) {
+    switch(type){
+#define DEFINE_CASE(x) case c10::ScalarType::x: return TorchTensorType##x;
+        DEFINE_TENSOR_TYPES(DEFINE_CASE)
+#undef DEFINE_CASE
+        default: return TorchTensorTypeUndefined;
+    }
+}
+
+@implementation TorchTensor {
+    std::shared_ptr<at::Tensor> _impl;
```

Comment:
I think `at::Tensor` / `at::IValue` / `torch::jit::script::Module` are all pimpl types with `intrusive_ptr<T>` as their underlying data structure (e.g. `at::Tensor b = a` would bump the refcount of `a`'s underlying `intrusive_ptr<TensorImpl>`). So we might not need `std::shared_ptr<at::Tensor> _impl` and can just use `at::Tensor _impl`.